### PR TITLE
multi-threaded version of freeipmi.plugin

### PR DESCRIPTION
**Pull request by ktsaou**
Fixes: #15325
Fixes: #15167

- [x] multi-threaded (a thread is collecting sensor data, another is collecting IPMI events, a third is updating netdata)
- [x] one chart per sensor
- [x] added pattern based component monitored
- [x] added sensor state chart (one per sensor) having 4 dimensions (`nominal`, `warning`, `critical` and `unknown`)

### CODE REVIEW

**Comment by thiagoftsm** [comment]
Hello @ktsaou ,

Your PR has the following warnings:

```sh
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'send_sensor_metrics_to_netdata':
collectors/freeipmi.plugin/freeipmi_plugin.c:1255:16: warning: unused variable 'sensors_states_chart_generated' [-Wunused-variable]
 1255 |     static int sensors_states_chart_generated = 0;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:906,
                 from /usr/include/malloc.h:25,
                 from ./libnetdata/libnetdata.h:73,
                 from collectors/freeipmi.plugin/freeipmi_plugin.c:21:
In function 'printf',
    inlined from 'send_sensor_metrics_to_netdata' at collectors/freeipmi.plugin/freeipmi_plugin.c:1379:25,
    inlined from 'main' at collectors/freeipmi.plugin/freeipmi_plugin.c:2094:9:
/usr/include/bits/stdio2.h:86:10: warning: 'context' may be used uninitialized [-Wmaybe-uninitialized]
   86 |   return __printf_chk (__USE_FORTIFY_LEVEL - 1, __fmt, __va_arg_pack ());
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'main':
collectors/freeipmi.plugin/freeipmi_plugin.c:1269:29: note: 'context' was declared here
 1269 |                 const char *context;
      |                             ^~~~~~~
In function 'printf',
    inlined from 'send_sensor_metrics_to_netdata' at collectors/freeipmi.plugin/freeipmi_plugin.c:1379:25,
    inlined from 'main' at collectors/freeipmi.plugin/freeipmi_plugin.c:2094:9:
/usr/include/bits/stdio2.h:86:10: warning: 'title' may be used uninitialized [-Wmaybe-uninitialized]
   86 |   return __printf_chk (__USE_FORTIFY_LEVEL - 1, __fmt, __va_arg_pack ());
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'main':
collectors/freeipmi.plugin/freeipmi_plugin.c:1270:29: note: 'title' was declared here
 1270 |                 const char *title;
      |                             ^~~~~
In function 'printf',
    inlined from 'send_sensor_metrics_to_netdata' at collectors/freeipmi.plugin/freeipmi_plugin.c:1379:25,
    inlined from 'main' at collectors/freeipmi.plugin/freeipmi_plugin.c:2094:9:
/usr/include/bits/stdio2.h:86:10: warning: 'units' may be used uninitialized [-Wmaybe-uninitialized]
   86 |   return __printf_chk (__USE_FORTIFY_LEVEL - 1, __fmt, __va_arg_pack ());
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'main':
collectors/freeipmi.plugin/freeipmi_plugin.c:1271:29: note: 'units' was declared here
 1271 |                 const char *units;
      |                             ^~~~~
In function 'printf',
    inlined from 'send_sensor_metrics_to_netdata' at collectors/freeipmi.plugin/freeipmi_plugin.c:1379:25,
    inlined from 'main' at collectors/freeipmi.plugin/freeipmi_plugin.c:2094:9:
/usr/include/bits/stdio2.h:86:10: warning: 'family' may be used uninitialized [-Wmaybe-uninitialized]
   86 |   return __printf_chk (__USE_FORTIFY_LEVEL - 1, __fmt, __va_arg_pack ());
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'main':
collectors/freeipmi.plugin/freeipmi_plugin.c:1272:29: note: 'family' was declared here
 1272 |                 const char *family;
      |                             ^~~~~~
```

Best regards!

**Comment by andrewm4894** [comment]
Any changes to docs required as part of this?

https://github.com/netdata/netdata/blob/master/collectors/freeipmi.plugin/README.md

**Comment by ktsaou** [comment]
@thiagoftsm the code has been updated significantly. Can you check again for errors?

**Comment by ktsaou** [comment]
@ilyam8 can you please take care of the docs change about freeipmi?

The key difference is that now we generate one chart per sensor (grouped by units on the new UI) and there is also a state chart per sensor that has alarms attached to it.

Other than that, now the plugin is faster, sensors and SEL events do not interfere to each other and the data collection frequencies of the two are independent. The min data collection frequency of sensors remains 5 seconds, but for SEL the min is now 60 seconds.

(btw, let's discuss, I think I have found a way to speed up SEL too)

**Comment by thiagoftsm** [comment]
> @thiagoftsm the code has been updated significantly. Can you check again for errors?

Of course, I am checking right now. :handshake:

File: collectors/freeipmi.plugin/freeipmi_plugin.c [hunk 1]
```diff
@@ -1118,486 +883,869 @@ _ipmimonitoring_sensors (struct ipmi_monitoring_ipmi_config *ipmi_config)
...
+
+    for(i = 0; i < excluded_status_record_ids_length; i++) {
+        if(excluded_status_record_ids[i] == record_id)
+            return 1;
+    }
+
+    return 0;
+}
+
+static void sensor_set_value(struct sensor *sn, void *sensor_reading, struct netdata_ipmi_state *state) {
```
Comment (thiagoftsm) [inline collectors/freeipmi.plugin/freeipmi_plugin.c:1409]: Hello @ktsaou ,

We only have a warning for this function:

 ```sh
collectors/freeipmi.plugin/freeipmi_plugin.c: In function 'sensor_set_value':
collectors/freeipmi.plugin/freeipmi_plugin.c:1409:98: warning: unused parameter 'state' [-Wunused-parameter]
 1409 | static void sensor_set_value(struct sensor *sn, void *sensor_reading, struct netdata_ipmi_state *state) {
      |                                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```

I am going to run coverity on PR right now.

Comment (thiagoftsm) [reply collectors/freeipmi.plugin/freeipmi_plugin.c]: This issue was fixed with last commit, and probably coverity report was also fixed, I am running again to confirm.

File: collectors/freeipmi.plugin/freeipmi_plugin.c [hunk 2]
```diff
@@ -645,959 +711,694 @@ static int excluded_status_record_ids_check(int record_id) {
...
+        spinlock_unlock(&t->spinlock);
+    }
 
-            if ((oem_data_len = ipmi_monitoring_sel_read_oem_data (ctx, oem_data, 1024)) < 0)
-            {
-                collector_error( "ipmi_monitoring_sel_read_oem_data(): %s",
-                                 ipmi_monitoring_ctx_errormsg (ctx));
-                goto cleanup;
-            }
+    return ptr;
```
Comment (thiagoftsm) [inline collectors/freeipmi.plugin/freeipmi_plugin.c:1192]: Coverity is reporting here:

```sh
*** CID 462562:  Control flow issues  (DEADCODE)
/collectors/freeipmi.plugin/freeipmi_plugin.c: 1192 in netdata_ipmi_collection_thread()
1186     
1187             spinlock_lock(&t->spinlock);
1188             t->state = tmp_state;
1189             spinlock_unlock(&t->spinlock);
1190         }
1191     
>>>     CID 462562:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "return ptr;".
1192         return ptr;
1193     }
1194     
1195     // ----------------------------------------------------------------------------
1196     // sending data to netdata
1197     
```

File: collectors/freeipmi.plugin/freeipmi_plugin.c [hunk 3]
```diff
@@ -148,414 +167,469 @@ _init_ipmi_config (struct ipmi_monitoring_ipmi_config *ipmi_config)
...
-            switch(sn->sensor_reading_type) {
-                case IPMI_MONITORING_SENSOR_READING_TYPE_UNSIGNED_INTEGER8_BOOL:
-                    printf("SET i%d_n%d_r%d = %u\n"
-                           , sn->sensor_number
-                           , sn->record_id
-                           , sn->sensor_reading_type
-                           , sn->sensor_reading.bool_value
-                    );
-                    break;
+    netdata_update_ipmi_sel_events_count(state, sel_count >= 0 ? sel_count : 0);
```
Comment (thiagoftsm) [inline collectors/freeipmi.plugin/freeipmi_plugin.c:537]: It also reports a deadcode here:

```sh
*** CID 462561:  Control flow issues  (DEADCODE)
/collectors/freeipmi.plugin/freeipmi_plugin.c: 537 in netdata_get_ipmi_sel_events_count()
531                 NULL)) < 0) {
532             collector_error( "ipmi_monitoring_sel_by_record_id(): %s",
533                              ipmi_monitoring_ctx_errormsg (ctx));
534             goto cleanup;
535         }
536     
>>>     CID 462561:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "0" inside this statement: "netdata_update_ipmi_sel_eve...".
537         netdata_update_ipmi_sel_events_count(state, sel_count >= 0 ? sel_count : 0);
538     
539         rv = 0;
540     
541     cleanup:
542         if (ctx)
```

File: collectors/freeipmi.plugin/freeipmi_plugin.c [hunk 4]
```diff
@@ -645,959 +711,694 @@ static int excluded_status_record_ids_check(int record_id) {
...
-            "Sensor Name",
-            "Sensor Number",
-            "Sensor Type",
-            "Sensor State",
-            "Sensor Reading",
-            "Sensor Units",
-            "Sensor Event/Reading Type Code",
-            "Sensor Event Bitmask",
-            "Sensor Event String");
-#endif // NETDATA_COMMENTED
```
Comment (thiagoftsm) [inline collectors/freeipmi.plugin/freeipmi_plugin.c:904]: And finally another one here:

```sh
*** CID 462563:  Control flow issues  (DEADCODE)
/collectors/freeipmi.plugin/freeipmi_plugin.c: 906 in netdata_update_ipmi_sensor_reading()
900     
901         if(excluded_state) {
902             if(state->debug)
903                 fprintf(stderr, "Sensor '%s' is excluded for status check, by excluded_status_record_ids_check()\n", sensor_name);
904         }
905     
>>>     CID 462563:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression """" inside this statement: "t.sensor_name = strdupz((se...".
906         struct sensor t = {
907                 .sensor_type = sensor_type,
908                 .sensor_state = sensor_state,
909                 .sensor_units = sensor_units,
910                 .sensor_reading_type = sensor_reading_type,
911                 .sensor_name = strdupz(sensor_name ? sensor_name : "")
```

Comment (thiagoftsm) [reply collectors/freeipmi.plugin/freeipmi_plugin.c:904]: I am going to test compilation on FreeBSD now, but we need to verify whether these reports as false positive or not.

**Comment by thiagoftsm** [comment]
Except for the bug fix I made few minutes ago (https://github.com/netdata/netdata/pull/15331), because the software we use to develop renamed more than we expected, everything is fine on FreeBSD

File: collectors/freeipmi.plugin/freeipmi_plugin.c (hunk 2 above)
Comment (ktsaou) [reply collectors/freeipmi.plugin/freeipmi_plugin.c:1192]: Yes, this is for compatibility. Ignore.

File: collectors/freeipmi.plugin/freeipmi_plugin.c (hunk 3 above)
Comment (ktsaou) [reply collectors/freeipmi.plugin/freeipmi_plugin.c:537]: Also for compatibility. Ignore.

File: collectors/freeipmi.plugin/freeipmi_plugin.c (hunk 4 above)
Comment (ktsaou) [reply collectors/freeipmi.plugin/freeipmi_plugin.c:904]: Also for compatibility. Ignore.

**Comment by ktsaou** [comment]
So, I understand this is good to go.
Waiting for @ilyam8 to commit docs (and possibly alerts changes)

**Comment by thiagoftsm** [comment]
> So, I understand this is good to go. Waiting for @ilyam8 to commit docs (and possibly alerts changes)

You are right, I will compile last commit to be sure we are not missing nothing an approve PR after this.

**Review by thiagoftsm — DISMISSED** [review]
Last compilation was successful and coverity was false positive, LGTM!

**Comment by ilyam8** [comment]
> Any changes to docs required as part of this?
> 
> [`master`/collectors/freeipmi.plugin/README.md](https://github.com/netdata/netdata/blob/master/collectors/freeipmi.plugin/README.md)

I updated the collected metrics section in the documentation. The readme file will be updated later when we add the metadata file.

**Review by ilyam8 — APPROVED** [review]

**Pull request merged**